lib/cdo/redshift_import.rb

```diff
@@ -0,0 +1,161 @@
+  # @param schemas [String] Array of Redshift schemas in which to complete data import.
+  def self.complete_table_import(schemas)
+    schemas.each do |schema|
+      temporary_import_tables(schema).each do |import_table|
+        target_table = import_table.partition(TEMP_TABLE_PREFIX).last
+        backup_table = BACKUP_TABLE_PREFIX + target_table
+
+        # Rename existing table to back it up.
+        rename_table(schema, target_table, backup_table)
+
+        # Make staging table the production table.
+        rename_table(schema, import_table, target_table)
+
+        # Rename the primary key from '_import_[foo]_primary' to '[foo]_primary' so that the next time the
+        # DMS Replication Task runs and creates the staging table it can create the primary key without being blocked
+        # by an existing primary key with that name.
+        # Also rename the backup table's primary key.
+        old_primary_key = RedshiftImport.primary_key(schema, backup_table)
+        if old_primary_key
+          rename_primary_key(
+            schema,
+            backup_table,
+            old_primary_key[:name],
+            BACKUP_TABLE_PREFIX + old_primary_key[:name],
+            old_primary_key[:columns]
+          )
+        end
+
+        primary_key = RedshiftImport.primary_key(schema, target_table)
+        if primary_key
+          rename_primary_key(
+            schema,
+            target_table,
+            primary_key[:name],
+            primary_key[:name].partition(TEMP_TABLE_PREFIX).last,
+            primary_key[:columns]
+          )
+        end
+
+        # Drop the old production table.
+        drop_table(schema, backup_table)
+      end
+    end
+  end
+
+  # Get names of tables in a Redshift schema that were loaded via DMS to stage imported data from the database.
+  # @param schema [String] The Redshift schema to search for database import staging tables.
+  # Returns Array of Redshift table names.
+  def self.temporary_import_tables(schema)
+    redshift_client = RedshiftClient.instance
+    query = <<~SQL
+      SET search_path TO #{schema};
+      SELECT DISTINCT t.tablename
+      FROM pg_table_def t LEFT JOIN pg_indexes i ON t.tablename = i.indexname
+      WHERE  i.indexname IS NULL -- Don't count primary keys, which appear in pg_table_def, as a table.
+        AND  t.schemaname ='#{schema}'
+        AND t.tablename LIKE '#{TEMP_TABLE_PREFIX}%';
+    SQL
+    redshift_client.exec(query).map {|row| row['tablename']}
+  end
+
+  # Get name and columns of primary key constraint for a specific table, if constraint exists.
+  # @param schema [String] Redshift schema.
+  # @param table [String] The Redshift table.
+  # Returns hash containing name of primary key and an array of the column names it constrains.
+  def self.primary_key(schema, table)
+    redshift_client = RedshiftClient.instance
+    primary_key_query = <<~SQL
+      SELECT co.conname AS constraint_name
+        ,    co.conkey AS constraint_column_ids
```

Comment:
This column has an integer array type, which Redshift does not support.  Looking at the Redshift documentation more closely, it may still be possible to join on these array values to the `pg_attribute` table.  I may try that, since it would eliminate the string manipulation and second query.

Reply:
Looks like that isn’t possible.  This query returns the error “operator does not exist smallint[] = oid”

``` sql
SELECT co.conname AS constraint_name
  ,    co.conkey AS constraint_column_ids
  ,    a.*
FROM   pg_catalog.pg_tables t
        INNER JOIN pg_catalog.pg_class cl ON (cl.relname = t.tablename AND cl.relkind = 'r') -- relkind 'r' is an ordinary table
        INNER JOIN pg_catalog.pg_constraint co ON (co.conrelid = cl.oid AND co.contype = 'p') -- contype 'p' is primary key
        INNER JOIN pg_catalog.pg_namespace n ON cl.relnamespace = n.oid
        INNER JOIN pg_catalog.pg_attribute a ON co.conkey = a.attrelid
WHERE  t.schemaname = 'dashboard_production'
  AND  t.tablename = 'user_levels'
  AND  n.nspname = 'dashboard_production';
```